### PR TITLE
Add support for cli and json styled SR Linux configs

### PR DESCRIPTION
**Pull request by n0shut**
This is a matching PR to support CLI and JSON-styled configs for SR Linux https://github.com/srl-labs/srl-controller/pull/37

The idea is that KNE calculates how the config file should be named (`config.cli` or `config.json`) based on the extension provided to the startup config file.

Then srl-controller loads the appropriate startup config file.

## Changes to resetter
Resetter now resets to startup config if it was provided, or to a default config (factory) if no startup config was provided (see inline comments)

## Changes to ConfigPush
Config push now works with CLI snippets expressed both in `set /` flat syntax as well as multiline CLI-styled blobs.

## Changes to examples

I have thus removed the JSON config example and used CLI format hereafter.

### CODE REVIEW

File: topo/node/srl/srl.go [hunk 1]
```diff
@@ -25,7 +26,9 @@ import (
 
 const (
 	scrapliPlatformName = "nokia_srl"
-	configResetCmd      = "load factory auto-commit"
+	// srl-controller v0.6.0+ creates a named checkpoint "initial" on node startup
+	// configuration reset is therefore done by reverting to this checkpoint
+	configResetCmd = "/tools system configuration checkpoint initial revert"
```
Comment (n0shut) [inline topo/node/srl/srl.go:31]: resetting is now done by reverting to a checkpoint named `initial`.
This checkpoint is created by srl-controller:v0.6.0+ and will capture the config state of a node once it has booted and applied startup config (if present)

File: examples/nokia/srlinux-services/2node-srl-ixr6-with-oc-services.pbtxt [hunk 2]
```diff
@@ -11,46 +11,43 @@ nodes: {
...
+        # SR Linux can accept both complete config files in JSON format, or partial/full config snippets in the CLI format.
+        # nodes are configured with a partial config snippet in CLI format that adds configuration relevant for this example lab.
+        file: "config.cli"
     }
     services:{
         key: 22
         value: {
             name: "ssh"
             inside: 22
-            outside: 22
```
Comment (n0shut) [inline examples/nokia/srlinux-services/2node-srl-ixr6-with-oc-services.pbtxt:21]: outside ports were removed, as keys already convey this info

**Comment by alexmasi** [comment]
I took a quick glance, the behavior is exactly what we would want to support (cli or json), simplified cfgs, reset to initial cfg commit.

Can you update the examples/multivendor/. topo + cfg as well as cloudbuild/vendors/. topo + cfg to match the new simplified cfgs?

Assigning to Paul to do an actual code review here

**Comment by n0shut** [comment]
Thanks @alexmasi 
added CLI-styled configs for cloudbuild and multivendor

File: topo/node/srl/srl.go [hunk 3]
```diff
@@ -25,7 +27,10 @@ import (
...
+	pushCfgFile    = "/home/admin/kne-push-config"
```
Comment (hellt) [inline topo/node/srl/srl.go:33]: configpush functionality now undergoes the following steps:

1. config from a file is read into a string, and all double quotes are escpaed
2. the escaped config is then transferred to the srlinux node using pty transport opened with scrapligo over `kubectl exec`.
3. the transfer happens via `echo "<config lines>" > pushCfgFile` command.
4. then a candidate is opened and pushed config is sourced via `source pushCfgFile`
5. It is then committed and persisted.

File: topo/node/srl/testdata/configpush_failure [hunk 4]
```diff
@@ -10,13 +10,21 @@ A:pod1# environment complete-on-space false
 A:pod1# info from state system app-management application mgmt_server state | grep running
                 state running
 --{ running }--[  ]--
-A:pod1# info from state system configuration commit 1 status | grep complete
-                status complete
+A:pod1# file cat /etc/opt/srlinux/devices/app_ephemeral.mgmt_server.ready_for_config
+loaded initial configuration
```
Comment (hellt) [inline topo/node/srl/testdata/configpush_failure:14]: in srlinux-scrapli package v0.6.0 we improved the way to check when the NOS is ready to accept configs. This is the new way

File: topo/node/srl/srl.go [hunk 5]
```diff
@@ -134,10 +139,12 @@ func (n *Node) GenerateSelfSigned(ctx context.Context) error {
...
 
-	cfg, err := io.ReadAll(r)
-	cfgs := string(cfg)
+	cfgBytes, err := io.ReadAll(r)
+	// replace quotes in the config with escaped quotes, so that we can echo this config
+	// via `echo` CLI commands.
+	cfg := strings.ReplaceAll(string(cfgBytes), `"`, `\"`)
 
-	log.V(1).Infof("config to push:\n%s", cfgs)
+	log.V(1).Infof("config to push:\n%s", cfg)
```
Comment (bormanp) [inline topo/node/srl/srl.go:147]: Shouldn't these lines be after the check for err?

File: topo/node/srl/srl.go [hunk 6]
```diff
@@ -238,12 +271,13 @@ func (n *Node) Create(ctx context.Context) error {
 }
 
 func (n *Node) Delete(ctx context.Context) error {
-	err := n.ControllerClient.Delete(ctx, &srlinuxv1.Srlinux{ObjectMeta: metav1.ObjectMeta{Name: n.Name()}})
+	err := n.ControllerClient.Delete(ctx, &srlinuxv1.Srlinux{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: n.GetNamespace(), Name: n.Name()}},
+	)
```
Comment (bormanp) [inline topo/node/srl/srl.go:277]: When breaking up the line you should match the {}'s. 

```
        err := n.ControllerClient.Delete(ctx, &srlinuxv1.Srlinux{
                ObjectMeta: metav1.ObjectMeta{
                        Namespace: n.GetNamespace(), Name: n.Name(),
                },
        })
```

**Review by bormanp — APPROVED** [review]
The Go changes are simply nits

File: topo/node/srl/srl.go (hunk 6 above)
Comment (n0shut) [reply topo/node/srl/srl.go]: thanks
Done in [9a561cd](https://github.com/openconfig/kne/pull/343/commits/9a561cd50b503c75366a246a4ff40e33ffb8bea3)

File: topo/node/srl/srl.go (hunk 5 above)
Comment (n0shut) [reply topo/node/srl/srl.go:151]: thanks @bormanp 
done in [4cb5267](https://github.com/openconfig/kne/pull/343/commits/4cb5267965a65c22fc0c75dd3769ce06b3caeea5)

**Comment by alexmasi** [comment]
Looks like one of those commits caused the go test to fail

**Comment by hellt** [comment]
@alexmasi looks like some metallb tests failed -- https://github.com/openconfig/kne/actions/runs/4582237788/jobs/8092328043?pr=343#step:4:643
I don't think this PR is related to it

**Comment by alexmasi** [comment]
I have no clue how that test could have failed, the context gets cancelled as expected. Anyway rerunning now to confirm its a flake. This PR doesn't directly modify that code so I'm even more confused, we haven't seen a flake like that before

**Review by alexmasi — APPROVED** [review]
carrying over borman's approval that was dismissed, assuming the presubmit passes I will merge

sorry for the hassle here

**Review by alexmasi — APPROVED** [review]

**Pull request merged**